sympy/combinatorics/fp_groups.py

```diff
@@ -0,0 +1,724 @@
+        obj._free_group = fr_grp
+        obj._relators = relators
+        obj.generators = obj._generators()
+        obj.dytype = type("FpGroupElement", (FpGroupElement,), {"group": obj})
+        return obj
+
+    @property
+    def free_group(self):
+        return self._free_group
+
+    def relators(self):
+        return tuple(self._relators)
+
+    def _generators(self):
+        """Returns the generators of the associated free group."""
+        return self.free_group.generators
+
+
+# sets the upper limit on the number of cosets generated during
+# Coset Enumeration. "M" from Derek Holt's. It is supposed to be
+# user definable.
+CosetTableDefaultMaxLimit = 4096000
+max_stack_size = 500
+
+class CosetTable(Basic):
+    # coset_table: Mathematically a coset table
+    #               represented using a list of lists
+    # alpha: Mathematically a coset (precisely, a live coset)
+    #       represented by an integer between i with 1 <= i <= n
+    #       α ∈ c
+    # x: Mathematically an element of "A" (set of generators and
+    #   their inverses), represented using "FpGroupElement"
+    # fp_grp: Finitely Presented Group with < X|R > as presentation.
+    # H: subgroup of fp_grp.
+    # NOTE: We start with H as being only a list of words in generators
+    #       of "fp_grp". Since `.subgroup` method has not been implemented.
+
+    """
+
+    References
+    ==========
+
+    [1] Holt, D., Eick, B., O'Brien, E.
+    "Handbook of Computational Group Theory"
+
+    [2] John J. Cannon; Lucien A. Dimino; George Havas; Jane M. Watson
+    Mathematics of Computation, Vol. 27, No. 123. (Jul., 1973), pp. 463-490.
+    "Implementation and Analysis of the Todd-Coxeter Algorithm"
+
+    """
+    def __init__(self, fp_grp, subgroup):
+        self.fp_group = fp_grp
+        self.subgroup = subgroup
+        self.p = [0]
+        self.A = list(chain.from_iterable((gen, gen**-1) \
+                for gen in self.fp_group.generators))
+        self.table = [[None]*len(self.A)]
+        self.A_dict = {x: self.A.index(x) for x in self.A}
+        self.A_dict_inv = {}
+        for x, index in self.A_dict.items():
+            if index % 2 == 0:
+                self.A_dict_inv[x] = self.A_dict[x] + 1
+            else:
+                self.A_dict_inv[x] = self.A_dict[x] - 1
+        self.deduction_stack = []
+
+    @property
+    def omega(self):
+        """Set of live cosets
+        """
+        return [coset for coset in range(len(self.p)) if self.p[coset] == coset]
+
+    @property
+    def n(self):
+        """The number `n` represents the largest number that has been used so
+        far for a live coset
+        """
+        return max(self.omega) + 1
+
+    # checks whether the Coset Table is complete or not
+    def is_complete(self):
+        return not None in flatten(self.table)
+
+    # Pg. 153
+    def define(self, alpha, x):
+        A = self.A
+        if self.n == CosetTableDefaultMaxLimit:
```

Comment:
Maybe `len(self.table)` should be taken instead of `self.n` or else this should be an inequality. It is conceivable that the table could grow bigger than the limit with the tail containing only non-live cosets, and then a new live `beta` would be added making `self.n > CosetTableDefaultMaxLimit`.